cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -1969,6 +2042,13 @@ private module Cached {
     // by a function. This allows data to flow 'in' through references returned by a modeled
     // function such as `operator[]`.
     reverseFlow(nodeFrom, nodeTo)
+    or
+    // models-as-data summarized flow
+    FlowSummaryImpl::Private::Steps::summaryThroughStepValue(nodeFrom, nodeTo, _)
+    or
+    FlowSummaryImpl::Private::Steps::summaryLocalStep(nodeFrom.(FlowSummaryNode).getSummaryNode(),
+      nodeTo.(FlowSummaryNode).getSummaryNode(), true)
+    // TODO: should these really be in the same place?
```

Comment:
No, they should not. This disjunct: `FlowSummaryImpl::Private::Steps::summaryThroughStepValue(nodeFrom, nodeTo, _)` should not be part of the `simpleLocalFlowStep` predicate. See the QLDoc on the predicate:
> NOTE: This step should not be used in global data-flow/taint-tracking, but may be useful to include in the exposed local data-flow/taint-tracking relations.

So in keeping with that QLDoc we should move this call into the definition of `localFlowStep` (and not in `simpleLocalFlowStep`)

Reply:
Hmm.  The implementation of `localFlowStep` is currently nothing but a call to `simpleLocalFlowStep` - so I can move this case out to `localFlowStep`, but it will be for aesthetic reasons not for any difference in behaviour.

Reply:
> NOTE: This step should not be used in global data-flow/taint-tracking, but may be useful to include in the exposed local data-flow/taint-tracking relations.

This is a frustrating comment as there's no explanation of the reasoning.  But in any case as far as I can tell both are in the CPP local flow step relation (`simpleLocalFlowStep`), and both ask to be (`summaryThroughStepValue` via this comment, `summaryLocalStep` via the predicate name).  My confusion is why they are distinct predicates at all.

Reply:
Sorry, I missed your responses here. Let me provide some additional information:

As you know, there are two ways of doing dataflow: global dataflow (using a dataflow configuration), and local dataflow (using `localFlowStep*(a, b)` or whatever).

Global dataflow uses the `simpleLocalFlowStep` predicate and `localFlowStep` is for use by queries that require local flow instead of global dataflow. Global dataflow has a bunch of features that's not present in local dataflow (for example field flow).

One thing that would be nice to have in local dataflow is flow through modeled functions even though that's not technically global flow.

For example, it may be desirable for _local_ flow to identify flow in an example such as:
```cpp
int f() {
  char buf[10];
  fgets(buf, 10, stdin);
  std::vector<const char*> v;
  v.push_back(buf);
  // 
  auto s = v.front();
  // use s here
}
```
But MaD is implemented by synthesizing a bunch of flow into and out of modeled functions. So this flow won't be included automatically in local flow. Since this is simply global flow, however, it _will_ automatically be part of global dataflow.

To fix this, we can extend the `localFlowStep` relation with these additional steps. And by only including this in `localFlowStep` this won't extend the `simpleLocalFlowStep` relation.

This is why `simpleLocalFlowStep` has this as part of its QLDoc:
> It may have less flow than the `localFlowStep` predicate.

Now, up until now that hasn't actually been the case since `simpleLocalFlowStep` was identical to `localFlowStep`.

Does that make sense?

Reply:
... after reading your comment and the predicates involved several times, yes, I think that makes sense.  We should have named these predicates much more clearly to indicate intent.

I've just pushed a commit that I believe fixes this situation for the MAD steps in both dataflow and taint flow.  I've also added a bit more QLDoc + comments to clarify how these various predicates fit together.  Please check my comments are accurate!